File: Data Structures & Algorithms/longest-increasing-subsequence/submission-1.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        cache = {}
        def dfs(i,j):
            if i == len(nums):
                return 0

            key = (i,j)
            if key in cache:
                return cache[key]

            res = dfs(i+1,j)

            if j == -1 or nums[i] > nums[j]:
                res = max(res, 1+dfs(i+1,i))
            
            cache[key] = res            
            return res
        
        return dfs(0,-1)

File: Data Structures & Algorithms/sum-of-two-integers/submission-1.cpp
class Solution {
public:
    int getSum(int a, int b) {
        int c = 0;
        int s = 0;
        int res = 0;

        for (int i=0; i<32; i++) {
            int bit_a = (a >> i) & 1;
            int bit_b = (b >> i) & 1;
            s = bit_a ^ bit_b ^ c;
            c = (bit_a & bit_b) | ((bit_a | bit_b) & c);
            res |= (s << i);
        }
        return res ;
    }
};
